File: csharp/Creational/02-AbstractFactory/JobDispatcher.cs
public class JobDispatcher
{
    private readonly IAlertNotifier   _alertNotifier;
    private readonly IEmailReceipt    _emailReceipt;
    private readonly IDashboardWidget _dashboardWidget;

    public JobDispatcher(IClientNotificationFactory factory)
    {
        _alertNotifier   = factory.CreateAlertNotifier();
        _emailReceipt    = factory.CreateEmailReceipt();
        _dashboardWidget = factory.CreateDashboardWidget();
    }

    public void DispatchJob(string jobId, string clientEmail)
    {
        Console.WriteLine($"\n--- Dispatching Job #{jobId} ---");
        _alertNotifier.SendAlert(jobId,   "Your job has been assigned to a handyman.");
        _emailReceipt.SendReceipt(jobId,  clientEmail);
        _dashboardWidget.Render(jobId,    "Assigned");
    }
}

File: typescript/Creational/02-AbstractFactory/JobDispatcher.ts
class JobDispatcher {
  private alertNotifier:   IAlertNotifier;
  private emailReceipt:    IEmailReceipt;
  private dashboardWidget: IDashboardWidget;

  constructor(factory: IClientNotificationFactory) {
    this.alertNotifier   = factory.createAlertNotifier();
    this.emailReceipt    = factory.createEmailReceipt();
    this.dashboardWidget = factory.createDashboardWidget();
  }

  dispatchJob(jobId: string, clientEmail: string): void {
    console.log(`\n--- Dispatching Job #${jobId} ---`);
    this.alertNotifier.sendAlert(jobId,   "Your job has been assigned to a handyman.");
    this.emailReceipt.sendReceipt(jobId,  clientEmail);
    this.dashboardWidget.render(jobId,    "Assigned");
  }
}